docs/proposals/provider-agnostic-dns-health-checks.md

```diff
@@ -0,0 +1,86 @@
+
+1. It’s tightly coupled with Route53. If other DNS providers are supported they must either provide a similar feature, or health checks will not be supported
+2. Lacks the ability to reach endpoints in private networks
+
+This document describes a proposal to extend the current health check implementation to overcome these shortfalls.
+
+### Goals
+
+* Ability to configure health checks in the DNSPolicy associated to a Gateway
+* Ability for the health checks to reach endpoints in private networks
+* DNS records are disabled when the associated health check fails
+* Current status of the defined health checks is visible to the end user
+* Transparently keep support for other health check providers like Route53
+
+## Proposal
+
+### Control plane
+
+Add a field in the health check section of the DNSPolicy to select the health
+check type to use
+
+```yaml
+healthCheck:
+  # ...
+  port: "..."
+  # ...
+  type: "Provider | Agent"
+```
+
+The presence of the `healthCheck` field in the DNSPolicy will affect the reconciliation
+of the endpoints when creating/updating DNSRecords:
+
+* For every endpoint that is generated, a health check is created based on the configuration in the DNSPolicy
+* Relevant `providerSpecific` fields are added to the endpoint depending on the
+  `type`:
+
+  * `Provider`: In the case of AWS, the `aws/health-check-id` field will be set with the ID of the health check in Route53
+  * `Agent`:
+    * A reference to the equivalent `DNSHealthCheckProbe` in the downstream cluster
+    will be set: `agent/probe-name`, `agent/probe-namespace`
+    * The health status of the `DNSHealthCheckProbe` will be propagated: `agent/health-status`
+
+
+### Data plane
+
+#### `DNSHealthCheckProbe` resource
+
+The DNSHealthCheckProbe resource configures the agent in the data plane cluster to perform the health checks against a local endpoint.
+
+```yaml
+apiVersion: kuadrant.io/v1alpha1
+kind: DNSHealthCheckProbe
+metadata:
+  name: example-probe
+spec:
+  port: "..."
+  host: “...”
+  ipAddress: "..."
+  path: "..."
+  protocol: "..."
+  interval: "..."
+status:
+  status: Healthy
+  lastCheck: "..."
+```
+
+The reconciliation of this resource results in the configuration of the health check agent, which targets the endpoint and updates the status. The status is propagated upstream to the providerSpecific status of the equivalent endpoint in the DNSRecord
+
+### Changes to current controllers
+
+In order to support this new feature, the following changes in the behaviour of the controllers are proposed.
+
+#### DNSPolicy controller
+
+Currently the reconciliation loop of this controller is in charge of creating health checks in the configured DNS provider (Route53 currently) based on the spec of the DNSPolicy, separately from the reconciliation of the DNSRecords. The proposed change is to reconcile health checks as the DNSRecords are created, using the health check
+configuration to alter the behaviour of the DNSRecord reconciliation.
```

Comment:
I am not getting if the DNSHealthCheckProbe needs to be synced to the spokes or if it stays in the control plane and is read by the agents on the spokes?

Reply:
It would be syncted to the spokes, and reconciled by the agent that runs in the spokes, then the status would be synced back to the control plane